[SwiftUI] Provide ability for clients to tell when WebView scroll geometry has changed

rr-codes opened the pull request:
#### 9240eac1e801f2fe5f1d0bb30ef1a7066a762b78
<pre>
[SwiftUI] Provide ability for clients to tell when WebView scroll geometry has changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=288242">https://bugs.webkit.org/show_bug.cgi?id=288242</a>
<a href="https://rdar.apple.com/145321354">rdar://145321354</a>

Reviewed by Aditya Keerthi.

Add a new `webViewOnScrollGeometryChange` view modifier to `WebView` to provide clients the ability
to respond to changes to various sizes and geometries that are modified by scrolling.

To facilitate this, layer tree transaction updates are propogated through a new `WKUIDelegateInternal`
protocol. The `WKUIDelegateAdapter` type now conforms to this new protocol, and passes the information
to `CocoaWebViewAdapter`, which then notifies the closures provided by the view modifier.

The `WKUIDelegateAdapter` type is made to more accurately reflect its conformance by conforming to `WKUIDelegatePrivate`.
However, this is still not _entirely_ accurate since it does also conform to `WKUIDelegateInternal`, however this
conformance cannot be statically stated since types imported via `internal import` can&apos;t appear in places that are outside
the project scope, which `WKUIDelegateAdapter` is.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::geometryDidChange):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKScrollGeometry.h: Copied from Source/WebKit/Modules/Internal/WebKitInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKScrollGeometry.mm: Copied from Source/WebKit/Modules/Internal/WebKitInternal.h.
(-[_WKScrollGeometry initWithContainerSize:contentInsets:contentOffset:contentSize:]):
* Source/WebKit/UIProcess/API/Swift/WKScrollGeometryAdapter.swift: Copied from Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift.
* Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift:
(WKUIDelegateAdapter._webView(_:geometryDidChange:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPageWebView.geometryDidChange(_:)):
(Delegate.geometryDidChange(_:)):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::geometryDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/_WebKit_SwiftUI/CocoaWebViewAdapter.swift:
(CocoaWebViewAdapter.onScrollGeometryChange):
(CocoaWebViewAdapter.currentScrollGeometry):
(CocoaWebViewAdapter.contentView):
(CocoaWebViewAdapter.findBarViewDidChangeHeight):
(CocoaWebViewAdapter.geometryDidChange(_:)):
* Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
* Source/WebKit/_WebKit_SwiftUI/SwiftUI+Extras.swift: Added.
* Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
(EnvironmentValues.webViewOnScrollGeometryChange):
(View.webViewContentBackground(_:)):
(FindContext.isPresented): Deleted.
* Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift: Copied from Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift.
(FindContext.isPresented):
* Source/WebKit/_WebKit_SwiftUI/WebViewRepresentable.swift:
(WebViewRepresentable.updatePlatformView(_:context:)):

Canonical link: <a href="https://commits.webkit.org/290966@main">https://commits.webkit.org/290966@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68755ac4723c0f5a3599b5f1771374442af0d6f0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91611 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11141 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/678 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96576 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19573 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70342 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94612 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50667 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41461 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98584 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/18766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/78842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/78575 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14508 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/18760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/18469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/21926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/20231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/SwiftUI+Extras.swift:
```diff
@@ -0,0 +1,10 @@
+
+extension ScrollGeometry {
+    init(_ geometry: WKScrollGeometryAdapter) {
+        self = ScrollGeometry(contentOffset: geometry.contentOffset, contentSize: geometry.contentSize, contentInsets: .init(), containerSize: geometry.containerSize)
```
need to fix contentInsets

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKScrollGeometry.mm:
```diff
@@ -0,0 +1,51 @@
+}
+
+@end
+
```
Stray newline.

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift:
```diff
@@ -35,7 +36,7 @@ private struct DefaultDialogPresenting: DialogPresenting {
 
 @MainActor
-final class WKUIDelegateAdapter: NSObject, WKUIDelegate {
+final class WKUIDelegateAdapter: NSObject, WKUIDelegatePrivate {
```
I think the commit message needs to explain why this isn't `WKUIDelegateInternal`.

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -359,6 +359,8 @@
     commitLayerAndScrollingTrees();
     webPageProxy->scrollingCoordinatorProxy()->didCommitLayerAndScrollingTrees();
 
+    // here
```
here

pxlcoder commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -13809,16 +13823,6 @@
-			);
-			path = protocol;
-			sourceTree = "<group>";
-		};
```
What's going on here?

rr-codes replied:
I have no idea, Xcode wanted this change. I can put it back if you want?

pxlcoder replied:
Maybe someone deleted the files and didn't remove from Xcode. Fine to leave as-is if so.

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/CocoaWebViewAdapter.swift:
```diff
@@ -221,11 +225,13 @@ class CocoaWebViewAdapter: PlatformView, PlatformTextSearching {
 #if os(macOS)
 extension CocoaWebViewAdapter: @preconcurrency NSTextFinderBarContainer {
-    func contentView() -> PlatformView? {
+    @_spi(CrossImportOverlay)
```
these changes are not needed

pxlcoder commented on Source/WebKit/Modules/Internal/WebKitInternal.h:
```diff
@@ -29,5 +29,7 @@
 #import "WKMaterialHostingSupport.h"
 #import "WKPreferencesInternal.h"
+#import "WKScrollGeometry.h"
+#import "WKUIDelegatePrivate.h"
```
I think it's better to import the Internal header here for consistency (even though the Swift code can't actually declare conformance to it).

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -2269,4 +2272,28 @@ static _WKXRSessionFeatureFlags toWKXRSessionFeatureFlags(const PlatformXR::Devi
+    if (![delegate respondsToSelector:@selector(_webView:geometryDidChange:)])
+        return;
+
+    [delegate _webView:uiDelegate->m_webView.get().get() geometryDidChange:geometry.get()];
```
I know you check for equality later on, in the Swift code, but it seems wrong to call the delegate at all if the geometry is unchanged. Can we move the equality check up here to avoid unnecessarily calling the delegate?

pxlcoder replied:
(or even before here, in WebPageProxy itself)

rr-codes replied:
> I know you check for equality later on, in the Swift code, but it seems wrong to call the delegate at all if the geometry is unchanged. Can we move the equality check up here to avoid unnecessarily calling the delegate?

yup, will fix. (I do have to leave the Swift code still though since it does a stricter equality check of the transformed value).

pxlcoder commented on Source/WebKit/_WebKit_SwiftUI/WebViewRepresentable.swift:
```diff
@@ -52,6 +51,8 @@ struct WebViewRepresentable {
         webView.allowsBackForwardNavigationGestures = environment.webViewAllowsBackForwardNavigationGestures.value != .disabled
         webView.allowsLinkPreview = environment.webViewAllowsLinkPreview.value != .disabled
 
+        let isOpaque = environment.webViewContentBackground != .hidden
```
Unrelated change?

rr-codes replied:
quite; thanks for catching!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2953,6 +2954,30 @@ - (_WKRectEdge)_fixedContainerEdges
 
+- (void)_updateScrollGeometryWithContentOffset:(CGPoint)contentOffset contentSize:(CGSize)contentSize
+{
+    CGSize containerSize = [self frame].size;
```
```suggestion
    CGSize containerSize = self.frame.size;
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2953,6 +2954,30 @@ - (_WKRectEdge)_fixedContainerEdges
+        return;
+
+    [uiDelegate _webView:self geometryDidChange:newScrollGeometry.get()];
+    _currentScrollGeometry = newScrollGeometry;
```
Can we swap the order of these lines? It's more idiomatic to update state prior to making delegate calls.

rr-codes committed:
[SwiftUI] Provide ability for clients to tell when WebView scroll geometry has changed
https://bugs.webkit.org/show_bug.cgi?id=288242
rdar://145321354

Reviewed by Aditya Keerthi.

Add a new `webViewOnScrollGeometryChange` view modifier to `WebView` to provide clients the ability
to respond to changes to various sizes and geometries that are modified by scrolling.

To facilitate this, layer tree transaction updates are propogated through a new `WKUIDelegateInternal`
protocol. The `WKUIDelegateAdapter` type now conforms to this new protocol, and passes the information
to `CocoaWebViewAdapter`, which then notifies the closures provided by the view modifier.

The `WKUIDelegateAdapter` type is made to more accurately reflect its conformance by conforming to `WKUIDelegatePrivate`.
However, this is still not _entirely_ accurate since it does also conform to `WKUIDelegateInternal`, however this
conformance cannot be statically stated since types imported via `internal import` can't appear in places that are outside
the project scope, which `WKUIDelegateAdapter` is.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::geometryDidChange):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKScrollGeometry.h: Copied from Source/WebKit/Modules/Internal/WebKitInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKScrollGeometry.mm: Copied from Source/WebKit/Modules/Internal/WebKitInternal.h.
(-[_WKScrollGeometry initWithContainerSize:contentInsets:contentOffset:contentSize:]):
* Source/WebKit/UIProcess/API/Swift/WKScrollGeometryAdapter.swift: Copied from Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift.
* Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift:
(WKUIDelegateAdapter._webView(_:geometryDidChange:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPageWebView.geometryDidChange(_:)):
(Delegate.geometryDidChange(_:)):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::geometryDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/_WebKit_SwiftUI/CocoaWebViewAdapter.swift:
(CocoaWebViewAdapter.onScrollGeometryChange):
(CocoaWebViewAdapter.currentScrollGeometry):
(CocoaWebViewAdapter.contentView):
(CocoaWebViewAdapter.findBarViewDidChangeHeight):
(CocoaWebViewAdapter.geometryDidChange(_:)):
* Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
* Source/WebKit/_WebKit_SwiftUI/SwiftUI+Extras.swift: Added.
* Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
(EnvironmentValues.webViewOnScrollGeometryChange):
(View.webViewContentBackground(_:)):
(FindContext.isPresented): Deleted.
* Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift: Copied from Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift.
(FindContext.isPresented):
* Source/WebKit/_WebKit_SwiftUI/WebViewRepresentable.swift:
(WebViewRepresentable.updatePlatformView(_:context:)):

Canonical link: https://commits.webkit.org/290966@main

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKScrollGeometry.h:
```diff
@@ -0,0 +1,54 @@
+#endif
+
+NS_SWIFT_SENDABLE
+@interface WKScrollGeometry : NSObject
```
Do you want an audit region around this header?

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKScrollGeometry.mm:
```diff
@@ -0,0 +1,67 @@
+    if (other == self)
+        return YES;
+
+    RetainPtr otherGeometry = dynamic_objc_cast<WKScrollGeometry>(other);
```
Is this just a WebKit-ism that you don’t use `-isKindOfClass:` here?

rr-codes replied:
ya; dynamic_objc_cast does both a class check and a cast (and nil if the class check fails)

axiixc commented on Source/WebKit/UIProcess/API/Swift/WKScrollGeometryAdapter.swift:
```diff
@@ -0,0 +1,56 @@
+
+@_spi(CrossImportOverlay)
+public struct WKScrollGeometryAdapter {
+    @_spi(CrossImportOverlay)
```
Do you need this repeated? I’d assume `public let` would be enough (and it’s not like a `public` field can be more visible than the containing type w.r.t. interface generation… I hope).

rr-codes replied:
good point, thanks!!!

axiixc commented on Source/WebKit/_WebKit_SwiftUI/CocoaWebViewAdapter.swift:
```diff
@@ -248,6 +252,29 @@ extension CocoaWebViewAdapter: WebPageWebView.Delegate {
+            return
+        }
+
+        onScrollGeometryChange.action(transformedOld, transformedNew)
```
Since this (presumably) kids off side effects, do you want it to occur before or after the ivar gets updated? Since you’re using `defer` it’ll happen after right now, but just wanted to check that that’s intentional.

axiixc commented on Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
```diff
@@ -34,4 +34,18 @@ func onNextMainRunLoop(do body: @escaping @MainActor () -> Void) {
+
+    init<E: Equatable>(_ value: E) {
+        self.value = value
+        self.equals = { ($0 as! E) == value }
```
The `as!` seems problematic here. As a type eraser, aren’t you _expecting_ to support mismatched values? This will crash instead of returning `false`.
